server/src/completions.ts

```diff
@@ -0,0 +1,432 @@
+  allowedTypes: string[],
+  kind: CompletionItemKind,
+): CompletionItem[] {
+  return allowedTypes.map((label) => ({ label, kind }))
+}
+
+function getSuggestionForBlockAttribute(
+  blockType: string,
+  document: TextDocument,
+  position: Position,
+): CompletionItem[] {
+  if (blockType !== 'model') {
+    return []
+  }
+
+  const symbolsBeforePosition = document.getText({
+    start: { line: position.line, character: position.character - 2 },
+    end: { line: position.line, character: position.character },
+  })
+  const symbolBeforePosition = symbolsBeforePosition.charAt(1)
+  // create deep copy
+  const suggestions: CompletionItem[] = klona(blockAttributes)
+
+  if (symbolsBeforePosition === ' @') {
+    // add @
+    for (const sugg of suggestions) {
+      sugg.label = '@' + sugg.label
+    }
+  } else if (/\s/.exec(symbolBeforePosition)) {
+    // add @@
+    for (const item of suggestions) {
+      item.label = '@@' + item.label
+    }
+  }
+
+  return suggestions
+}
+
+function getSuggestionForFieldAttribute(
+  blockType: string,
+  currentLine: string,
+  document: TextDocument,
+  position: Position,
+): CompletionItem[] {
+  if (blockType !== 'model' && blockType !== 'type_alias') {
+    return []
+  }
+  // create deep copy
+  let suggestions: CompletionItem[] = klona(fieldAttributes)
+
+  if (!currentLine.includes('Int')) {
+    // id not allowed
+    suggestions = suggestions.filter((sugg) => sugg.label !== 'id')
+  }
+  const symbolBeforePosition = document.getText({
+    start: { line: position.line, character: position.character - 1 },
+    end: { line: position.line, character: position.character },
+  })
+  if (symbolBeforePosition !== '@') {
+    // add @
+    for (const sugg of suggestions) {
+      sugg.label = '@' + sugg.label
+    }
+  }
+
+  return suggestions
+}
+
+export function getSuggestionsForAttributes(
+  blockType: string,
+  position: Position,
+  document: TextDocument,
+  currentLine: string,
+): CompletionList | undefined {
+  // check if position is after field name and type!
+  const foundIndex = currentLine.indexOf('@')
+  if (foundIndex === -1) {
+    return
+  }
+  const wordsTillSymbol = currentLine.slice(0, foundIndex)
+  const matchArray = wordsTillSymbol.match(/\s+/g)
+  if (!matchArray || matchArray.length < 2) {
+    return
+  }
+
+  const suggestions: CompletionItem[] = getSuggestionForFieldAttribute(
+    blockType,
+    currentLine,
+    document,
+    position,
+  )
+
+  return {
+    items: suggestions,
+    isIncomplete: false,
+  }
+}
+
+export function getAllRelationNamesWithoutAst(
+  lines: Array<string>,
+  currentModelStartLine: number,
+): Array<string> {
+  const modelNames: Array<string> = []
+  for (const [key, item] of lines.entries()) {
+    if (currentModelStartLine === key) {
+      // do not suggest the model name we are currently in
+      continue
+    }
+    if (
+      (item.includes('model') || item.includes('enum')) &&
+      item.includes('{')
```

Comment:
This relies on Prisma fmt to work and being used. It won't support this variant: 

```
model User 
{
```

We might want to use a parser for this one (`prismafile`?). There could be multiple reasons why a team don't use Prisma fmt (like conventions, failure to know that format on save is an option etc).

Reply:
`Prismafile` is not able to parse error-tolerant, also this will soon be unsupported as well, since this functionality will be replaced soon for the SDK. 

It's not supposed to work on this variant, because then we would need to define what's possible. E.g. allow writing the whole schema in one line, or having mulitple new lines in between the modelName and `{`, ... . Not sure if this is worth putting the time into. @janpio?

Reply:
Difficult question. Auto completion should of course already work before formatting was applied, but if I read your comment right this would be a lot more complicated and complex right now @carmenberndt, correct?

Reply:
yes, this would be quite easy to implement once we have a solution for the error-tolerant-parser. Then I can simply check where a model begins and ends and so on, right now this is quite complex since I am only working with strings here.